### DISCUSSION TITLE
Add Text To Image from Metadata

### DISCUSSION THREAD

**Discussion by GoremanX** [ImageMagick/ImageMagick#7274] [Help] [answered]
### ImageMagick version

7.1.1

### Operating system, version and so on

Windows 10

### Description

I'd like to add text to the bottom of an image in a specific font type and size, horizontally centered and X pixels from the bottom. The text is the image's Title metadata. I've attached an example result, but I've got hundreds of these to do. Any help in figuring out the command to accomplish this would be greatly appreciated. Bonus points if I can turn any space in the text into a line break

[hidden]

**Comment by fmw42** [comment]
Post the input image so we can see where the text comes from in the meta data.

**Comment by GoremanX** [comment]
It's much more complex than I'm implying. I've been wading through the documentation and trying countless iterations to get it working, and I'm closer than I was but still not there.

My current command looks like this:
`magick polaroid-frame.png in.jpg -geometry +182+304 -compose DstOver -composite -gravity center -size 1550x -font Eraser -pointsize 150 -annotate -0+950 caption:"%[EXIF:Title]" pnged\out.png`

Source files are as follows:

[hidden]

![polaroid-frame](https://github.com/ImageMagick/ImageMagick/assets/876310/e489c6b6-ec00-4020-aca1-ab75be0880b9)

The "Title" tag in the JPG contains the student's name. I want to display the name under his photo, centered, and preferrably wrapping within the width that his photo occupies. I'm getting VERY confused with all the references I'm finding, many of them just outright don't do what they say they do, or they're references for ImageMagick 6 that don't work in 7. With my command line, wrapping doesn't work at all, the word "caption:" keeps showing up before the text, and the Title tag doesn't appear to be found at all.

**Reply by GoremanX** [reply]
And just to show I'm not on crack:

`exiftool in.jpg -title`
`Title                           : Cooper Gordon`

**Comment by fmw42** [comment]
Do not use both -annotate and caption:.  One or the other.  -annotate does not wrap, but writes directly to the image.  Caption wraps but you have to create a new image (via caption) and composite that over the background image.


Perhaps I am missing it, but I do not see the EXIF:title in the IM list of supported tags. See https://imagemagick.org/script/escape.php. 

When I do

exiftool -s -ee -g1 -u -D in.jpg

I see the title but it is listed under XMP tags.  So it is hidden in the XMP profile section, not the EXIF section. I do not know how to extract from the XMP section in Imagemagick. You could use EXIFTOOL in a script to get the title and pass that as a variable to Imagemagick.

So I have used the Artist tag for now.  Does this do what you want?


Unix syntax:

```
magick polaroid-frame.png \
\( in.jpg -set option:title "%[EXIF:Artist]" \) \
-geometry +182+304 -compose over -composite \
\( -background none -size 1540x400 -pointsize 150 \
-fill black -font /Users/fred/fonts/Eraser.ttf \
-gravity center caption:"%[Title]" \) \
-gravity northwest -geometry +210+2315 -compose over -composite \
out.png

```

Windows Syntax (note, you may need to double the % to %%):

```
magick polaroid-frame.png ^
( in.jpg -set option:title "%[EXIF:Artist]" ) ^
-geometry +182+304 -compose over -composite ^
( -background none -size 1540x400 -pointsize 150 ^
-fill black -font /Users/fred/fonts/Eraser.ttf ^
-gravity center caption:"%[Title]" ) ^
-gravity northwest -geometry +210+2315 -compose over -composite ^
out.png
```

![out](https://github.com/ImageMagick/ImageMagick/assets/13458218/71ae50e2-82ce-4202-9ab4-cdf5f39c37ad)

**Reply by GoremanX** [reply]
that works exactly the way I was hoping for, THANK YOU! I could probably run exiftool to copy the title tag to the artist tag on every file, it's not like I need to know who took the picture (me!)

**Reply by fmw42** [reply, marked as answer]
Here is how to get the Title tag from the header using EXIFTOOL and pipe it to Imagemagick with a little editing to remove the leading spaces and the trailing new line.  No need to save it to a variable.

Windows Syntax:

```
exiftool in.jpg -title | cut -d: -f2 | sed 's/^[ ]*//' | tr -d '^n' |
magick polaroid-frame.png in.jpg ^
-geometry +182+304 -compose over -composite ^
( -background none -size 1540x400 -pointsize 150 ^
-fill black -font /Users/fred/fonts/Eraser.ttf ^
-gravity center caption:"@-" ) ^
-gravity northwest -geometry +210+2315 -compose over -composite ^
out2.png

```

**Reply by GoremanX** [reply]
Pretty sure cut and sed aren't on Windows, sadly. Which is infuriating because I'd love to be able to convert the students' names to all uppercase, but I'm not at all familiar with a Windows-way to accomplish that on the command line. Having said that, I could just move all these files to a linux box and run the command there

**Reply by fmw42** [reply]
Yes, you can get them for Windows if you do not already have them.  Check with Google.

**Reply by fmw42** [reply]
See for example https://www.onlogic.com/company/io-hub/how-to-enable-bash-for-windows-10-and-11/ and https://www.laptopmag.com/articles/use-bash-shell-windows-10 and https://superuser.com/questions/608106/how-can-i-use-a-bash-like-shell-on-windows and https://www.geeksforgeeks.org/use-bash-shell-natively-windows-10/ and many others.  I searched Google for "bash tools for windows"

**Reply by GoremanX** [reply]
One slight hiccup, the command appears to omit the letter "n" when that's the last letter of a word. I changed it to `tr ' ' '\n'` and that seemed to resolve the issue. I also added another pipe ` | tr [:lower:] [:upper:]` to make the names all uppercase, and changed the path to the font file. Everything works wonderfully now, thank you SO MUCH for your help.

**Comment by snibgo** [comment]
> Pretty sure cut and sed aren't on Windows, sadly. 

cut, sed and tr are not provided *with* Windows, but they are available *on* Windows. I use versions of these utilities from Cygwin that can be run either from Windows CMD shell, or BAT scripts, or (Cygwin) bash.

**Reply by GoremanX** [reply]
Thank you, that worked very well
